0013 - Use TypeScript on the project (#19)
PrestaShop/ADR: 0013-use-typescript-on-the-project.md

# 13. Use TypeScript on the project

Date: 2021-02-25

## Status

> In discussion

Reviewer: ```suggestion
Accepted
```

## Context

While the whole JavaScript community welcome TypeScript easily, our project JS is not typed at all. In order to improve the maintainability of the project and use latest TC39 features, a good idea would be to use TypeScript.

What would benefit the project of using it :

> - Detect bugs before sending PRs, TypeScript users says that globally, it permit to detect around 15% of bugs that you would detect by testing.

Reviewer 2: ```suggestion
- Detect bugs before sending PRs, TypeScript users say that globally, it allows to detect around 15% of bugs that you would detect by testing.
```

- Use latest features such as Optional Chaining, Tuples and Records... really early.
- Types are increasing the quality of the project, because we would be able to detect dangerous changes, related bugs... If we use it on PHP side, why don't we use types while using JS ?

> - Vue 3 offered a new API : Composition API, this one is pretty easy to use with TypeScript as it's mainly functional programming instead of opiniated APIs of Vue, that would be a good moove to preshot the Vue update in the BO.

Reviewer 2: ```suggestion
- Vue 3 offered a new API : Composition API, this one is pretty easy to use with TypeScript as it's mainly functional programming instead of opiniated APIs of Vue, that would be a good move to preshot the Vue update in the BO.
```
